### DISCUSSION TITLE
Bundling for AWS Lambda deployment

### DISCUSSION THREAD

**Discussion by ygyg70** [TypedProject/ts-express-decorators#2624] [Q&A]
Before creating my own build configuration ([probably something like](https://docs.aws.amazon.com/cdk/api/v1/docs/aws-lambda-nodejs-readme.html#local-bundling)), I wanted to ask if there is already something out there I can use.

(or perhaps someone is kind enough to share their approach)

Any help appreciated!

**Comment by Romakita** [comment]
Hello @ygyg70 
@aws-cdk/aws-lambda-nodejs use esbuild which is not compatible with Ts legacyDecorator / emitDecoratorMetadata. You have to use this configuration:

```tsc
new lambda.NodejsFunction(this, 'my-handler', {
  bundling: {
    preCompilation: true,
  },
});
```

Has I explained on slack:
> use the complete Ts.ED framework is only useful for the local development (Having express server + swagger, will help dev to run code locally).
> But you have to deploy only the lambda in production

Your approach seems to be different from the Serverless tutorial https://tsed.io/tutorials/serverless.html#configuration
Using Serverless, the code embed the full Express Server (if you use the http version) or a light server (based on light router provided by fastify). But the problem of this approach (from my point of view) is that it combines the routing done by the Serverless framework and also goes through the Express/Koa/Fastify light router.

This is clearly not ideal and efficient.

Your proposition seems to be more interesting but there is not example for that.

The main idea is to have on handler/aws lambda that invoke the Ts.ED DI (https://github.com/tsedio/tsed/tree/production/packages/di):

**my-handler.ts**
```ts
import {InjectorService, attachLogger, DIContext} from "@tsed/di";
import {$log} from "@tsed/logger";
import {MyService} from "../usecases/MyService";

async function bootstrap(Service) {
  const injector = new InjectorService();

  // configure the default logger
  attachLogger(injector, $log);

  // Load all providers registered via @Injectable decorator
  await injector.load();

  // emit event to trigger actions for third parties modules
  await injector.emit("$onReady");

  return injector
}

const promise = bootstrap(); // to pre build Container before lambda handler invokation

export default  async function handler(awsArgs) { // need to be completed
   const injector = await promise

   const ctx = new DIContext({
      id: /* aws request.id || */ uuid.v4(),
      injector,
      logger: injector.logger
   })

   ctx.set("aws_args", awsArgs) // customise that to you convenient
   
   const myService = injector.get<MyService>(MyService);
   
   // we need to map aws Args to keep our MyService agnostic from AWS /!\ important
   // and to reuse MyService inside a Ts.ED controller (clean architecture)
   
   const {value, other} = mapAwsArg(AwsArgs);
 
   try {
      const response = await runInContext(ctx, () => myService.doSomething(value, other));
   
      return mapResponseToAws(response)  
   } catch(er) { 
      // handle response
   }
} 
```

Tell me if you have some question ;)
See you

**Reply by Romakita** [reply]
I encourage you to create a repository example that you can share here if it's necessary ;)

**Reply by ygyg70** [reply]
I really like the concepts Ts.ED is implementing, specifically usage of decorators, DI, aspects, full support of OpenAPI and best best practice architecture. 
I would like to establish Ts.ED as the framework of choice for Node microservices at my company. For that, there is a need to support scale to zero AWS deployment with few seconds cold start, and the only one currently approved is Lambda.
I am perfectly fine with the compromise of removing the express/koa layer, followed the tutorial here [https://tsed.io/tutorials/serverless.html](https://tsed.io/tutorials/serverless.html#configuration) and got it to work with a modification - ```export = platform.callbacks()``` didn't work for me, changed to ```export const handler= platform.handler();```

What I believe is still missing for the solution to be elegant enough:

- Automation to generate specs, deploy to an API gateway and connect to the Lambda, for dev, test and prod (dev/test may not be needed if the abstraction is complete enough so no defects are introduced when the API gateway is introduced)
- Tree-shaking build for creation of a compact deployment artifact - some may propose layers but they don't solve the code size aspects and break the concept of atomic builds

I am happy to spend some time on it and share results, but it ends up being large effort I will unfortunately have to bail.

I may be able to get some budget approved for a consultant to implementing the missing pieces.

I was also looking at Elastic Beanstalk with scale to zero, but our AWS contact recommended not using it..

**Reply by Romakita** [reply]
I totally forgot that :

```ts
import {PlatformServerless} from "@tsed/serverless";
import {TimeslotsLambda} from "./lambda/TimeslotsLambda";

const platform = PlatformServerless.bootstrap({
  lambda: [TimeslotsLambda]
  // put your Application configuration here
});

// then export the lambda
export = platform.callbacks(); // in my case I had to use this notation
// but it should works with export default.
export default platform.callbacks()
```
You can use that to build your handler. Here the TimeslotsLambda expose a CRUD and use a light router under the hood. 
You can by-pass the router by using `callback()`

```ts
const handler = PlatformServerless.factory({
  lambda: [TimeslotsLambda]
});

export default handler.callback(TimeslotsLambda, "getTimeslot") // the export must be adapted depending on the tsconfig / bundler options
```

**Reply by Romakita** [reply]
> Automation to generate specs, deploy to an API gateway and connect to the Lambda, for dev, test and prod (dev/test may not be needed if the abstraction is complete enough so no defects are introduced when the API gateway is introduced)
 
 Generate spec from the development server is possible through this plugin => https://www.npmjs.com/package/@tsed/cli-generate-swagger
    
> Tree-shaking build for creation of a compact deployment artifact - some may propose layers but they don't solve the code size aspects and break the concept of atomic builds

It's possible if you don't use barrelsby and use only direct ESM import. For example if you import only the LambdaController in your aws handler file, only file depending (Service/Module/etc...) on LambdaController will be shipped :)

> I may be able to get some budget approved for a consultant to implementing the missing pieces.

Great idea ;). You can maybe thinking about becomes Ts.ED sponsors (or select a one-time sponsor tiers) to having a dedicated private channel with me + custom development (See https://github.com/sponsors/Romakita)

See you

**Reply by ygyg70** [reply]
Hi @Romakita,
I suspect we may not be using the same versions here.. (I am using latest Ts.ED version):
 ```PlatformServerless``` doesn't have a method ```factory```
If I use ```export = platform.callbacks()``` or ```export default platform.callbacks()``` I get this error:
```Lambda handler "dist/handler" does not point function property```
You can find my example here: https://github.com/ygyg70/tsed-aws-lambda
Is there a repo where I can find your example?
Thanks!

**Reply by Romakita** [reply]
Yes you can find a short example here:
https://github.com/tsedio/tsed/tree/a6322b7ba5d7fd298d596b58e031b490d70c51ae/packages/platform/platform-serverless-http/test/integration/aws-basic

Currently, I haven't build a complete repository. And the example is probably outdated.

**Reply by ygyg70** [reply]
Thanks!
I am still working on creating a project that includes the following:

- Local dev/debug/test that includes generation of specs, deploy to an emulated API gateway that is running in a container, deploy to an emulated Lambda running on a container
- Packaging for efficient Lambda deployment
- Local execution of e2e tests and automated test (SchemaThesis) using the above topology
- If we see no discrepancy between server behavior (no Lambda, no API gateway) and serverless topology, we can drop dev/test of that and keep it just as a cloud deployment thing

Need to choose between SAM and paid version of LocalStack, SAM vs Terraform/other IaC, Serverless framework etc.
I am also looking at NestJS. As much as I like Ts.ED, NestJS seems to have a much bigger community/momentum, which is an important factor for enterprise adoption.

**Reply by Romakita** [reply]
Hello @ygyg70 
I understood your point to choose NestJS instead Ts.ED. 
Some points to convince your to choose Ts.ED

His learning Curve is much better than NestJS. This is a positive point that comes up the most from companies and developers who have chosen the framework.

I prefer to encourage the community to gives answer but I'm always available to give you answer. So in most time you have a answers during my working day.

I take a time to create open source repository to gives basic examples between two frameworks. 

Sometimes I even take the time to review open-source project code based on Ts.ED and which is used for commercial purposes.

But for Enterprise projects, I propose a sponsors tiers. Isn't enough but the gain is interesting for the company and me:
- You can have a private slack channel hosted on Ts.ED or directly to your slack.
- I can work for you to develop a base project

> Some companies choose to become premium sponsors and benefit from privileged access to ask me very specific questions on which I provide them with technical and personalized advice.

Right now, I understand that you need to benchmark each framework to choose the most valuable for you ;)

I can propose you to take a time to review your open source repository and try to gives you some feedback.
https://github.com/ygyg70/tsed-aws-lambda (or maybe another one if you want)
 
It's the least I can do to convince you ;)

See you

**Reply by ygyg70** [reply]
Going through your changes and trying to understand. Node, ES, TS, module management and tooling are quite a chore to master..
Looking at cold start benchmarks of https://github.com/awslabs/llrt, perhaps keeping it simple and using https://github.com/awslabs/aws-lambda-web-adapter and full deployment with Express/Koa is a better direction..

**Reply by Romakita** [reply]
XD. I understand that the example repo is a little obscure (it lacks clarification).

But your requests were to have a project taking advantage of tree shaking, executable/testable locally and which generates swagger.

That's exactly what I did :) and more:
- I added bundler compatible with Serverless, so us `yarn start:lambda`, you can compile and run lambda in offline mode (thanks Serverless)
- It's a scalable architecture (thanks mono repo).
- It start fast rather than an express server (cold start is supported).
- You can use find-my-way router to expose one controller or more, or create optimized handler without router. 
- You can develop your code locally using the Express framework (@tsed/platform-express).
- A good DX - can be improved - (your devs will thanks you)
  - Ts isn't really complicated to learn

> Perhaps keeping it simple and using https://github.com/awslabs/aws-lambda-web-adapter and full deployment with Express/Koa is a better direction..

This is exactly my first point when you ask me ;). 

But even if you're doing pure JS, don't put an express server behind it. Uses Serverless + JS handlers. You will have everything you need to structure your project and have the performance.

Note: I already tried a similar tools like aws-lambda-web-adapter ([packages/platform/platform-serverless-http/src/builder/PlatformServerlessHttp.ts](https://github.com/tsedio/tsed/blob/166dedefdf6e9531e17b93d7e2d10bc15546d90c/packages/platform/platform-serverless-http/src/builder/PlatformServerlessHttp.ts) and the problem isn't the cold start (not really) but the fact that the developers think they are in a classic Node.js Express application when it is completely wrong! 

It's like telling a React developer that React-native is the same ;)

Because a lambda is not written and should not be written like a Rest application. 

There are use cases where Lambda + Express will work: 
- Image Resizing
- S3 consumption and manipulation

But in this case, use Serveless configuration file is more interesting.

## Why Serverless ?

Because it's a framework compatible with many cloud provider. It's really an advantage to keep your code portable to another cloud provider (if you do the right division at the code level (Clean archi, DDD, etc...).

Also other points:
- many plugins to build and deploy (you can generate swagger)
- offline mode to emulate lambda 🤩  
- offline dynamo DB mode
- Embed router declaration and fine grained endpoint configuration. All that is converted to AWS using cloud formation.


--- 

You'll have understood but my main point is do not use Express/Koa for your project it will not bring you anything, if you use the right tools to emulate your lambdas locally. ;)

See you guy and thanks to have give me the opportunity to create a good tsed-aws-lambda project example

**Reply by ygyg70** [reply]
All this information and work are very useful :)
To clarify - when you say "Serverless + JS handlers" You mean the Lambda example that are in the repo above (my assumption), or something different?
Question - is schema validation supported with ESBuild Lambda? I added a Max constraint to ```TimeslotsController.getTimeslotById``` and it doesn't seem to work in Lambda mode (works in dev mode):
https://github.com/ygyg70/tsed-aws-lambda/tree/Add-path-param-with-validation
Assuming that if it is not supported by the app, it could be supported by the API gateway
